File: slp/backend-dotnet/Features/Explanation/ExplanationService.cs
using System.Text.Json;

namespace backend_dotnet.Features.Explanation;

public class ExplanationService : IExplanationService
{
    private readonly IExplanationRepository _repo;
    private readonly Source.ISourceRepository _sourceRepo;

    public ExplanationService(IExplanationRepository repo, Source.ISourceRepository sourceRepo)
    {
        _repo = repo;
        _sourceRepo = sourceRepo;
    }

    public async Task<IEnumerable<ExplanationDto>> GetBySourceAsync(int sourceId, int userId)
    {
        // Verify source belongs to user
        var source = await _sourceRepo.GetByIdAsync(sourceId);
        if (source == null || source.UserId != userId)
            return Enumerable.Empty<ExplanationDto>();

        var explanations = await _repo.GetBySourceIdAsync(sourceId, userId);
        return explanations.Select(MapToDto);
    }

    public async Task<ExplanationDto> CreateAsync(int userId, CreateExplanationRequest request)
    {
        var textRangeJson = JsonSerializer.Serialize(request.TextRange);

        var entity = new Explanation
        {
            UserId = userId,
            SourceId = request.SourceId,
            TextRangeJson = textRangeJson,
            Content = request.Content,
            AuthorType = "user",
            Editable = true,
            CreatedAt = DateTime.UtcNow
        };

        var created = await _repo.CreateAsync(entity);
        return MapToDto(created);
    }

    public async Task<ExplanationDto?> UpdateAsync(int id, int userId, UpdateExplanationRequest request)
    {
        var entity = await _repo.GetByIdAsync(id);
        if (entity == null) return null;

        // Only the owner can edit; system explanations are not editable
        if (entity.UserId != userId || !entity.Editable) return null;

        entity.Content = request.Content;
        await _repo.UpdateAsync(entity);
        return MapToDto(entity);
    }

    public async Task<bool> DeleteAsync(int id, int userId)
    {
        var entity = await _repo.GetByIdAsync(id);
        if (entity == null) return false;
        if (entity.UserId != userId) return false;

        await _repo.DeleteAsync(id);
        return true;
    }

    private static ExplanationDto MapToDto(Explanation e)
    {
        object? textRange = null;
        try { textRange = JsonSerializer.Deserialize<object>(e.TextRangeJson); } catch { }

        return new ExplanationDto
        {
            Id = e.Id,
            UserId = e.UserId,
            SourceId = e.SourceId,
            TextRange = textRange,
            Content = e.Content,
            AuthorType = e.AuthorType,
            Editable = e.Editable,
            CreatedAt = e.CreatedAt,
            UpdatedAt = e.UpdatedAt
        };
    }
}

File: slp/backend-nestjs/src/modules/explanation/explanation.service.ts
import { Injectable } from "@nestjs/common";
import { ExplanationRepository } from "./explanation.repository";
import { SourceRepository } from "../source/source.repository";
import { Explanation } from "./explanation.entity";
import { CreateExplanationRequest } from "./dto/create-explanation-request.dto";
import { ExplanationDto } from "./dto/explanation.dto";
import { UpdateExplanationRequest } from "./dto/update-explanation-request.dto";

@Injectable()
export class ExplanationService {
  constructor(
    private readonly explanationRepo: ExplanationRepository,
    private readonly sourceRepo: SourceRepository,
  ) {}

  private mapToDto(entity: Explanation): ExplanationDto {
    let textRange: any;
    try {
      textRange = JSON.parse(entity.textRangeJson);
    } catch {
      textRange = null;
    }
    return {
      id: entity.id,
      userId: entity.userId ?? undefined,
      sourceId: entity.sourceId,
      textRange,
      content: entity.content,
      authorType: entity.authorType,
      editable: entity.editable,
      createdAt: entity.createdAt,
      updatedAt: entity.updatedAt ?? undefined,
    };
  }

  async getBySource(
    sourceId: number,
    userId: number,
  ): Promise<ExplanationDto[]> {
    // Verify source belongs to user
    const source = await this.sourceRepo.getById(sourceId);
    if (!source || source.userId !== userId) {
      return [];
    }

    const explanations = await this.explanationRepo.getBySourceId(
      sourceId,
      userId,
    );
    return explanations.map(this.mapToDto);
  }

  async create(
    userId: number,
    request: CreateExplanationRequest,
  ): Promise<ExplanationDto> {
    const textRangeJson = JSON.stringify(request.textRange ?? {});

    const entity = new Explanation();
    entity.userId = userId;
    entity.sourceId = request.sourceId;
    entity.textRangeJson = textRangeJson;
    entity.content = request.content;
    entity.authorType = "user";
    entity.editable = true;

    const created = await this.explanationRepo.create(entity);
    return this.mapToDto(created);
  }

  async update(
    id: number,
    userId: number,
    request: UpdateExplanationRequest,
  ): Promise<ExplanationDto | null> {
    const entity = await this.explanationRepo.getById(id);
    if (!entity) return null;

    // Only owner can edit, and only if editable
    if (entity.userId !== userId || !entity.editable) return null;

    entity.content = request.content;
    const updated = await this.explanationRepo.update(entity);
    return this.mapToDto(updated);
  }

  async delete(id: number, userId: number): Promise<boolean> {
    const entity = await this.explanationRepo.getById(id);
    if (!entity) return false;
    if (entity.userId !== userId) return false;

    await this.explanationRepo.delete(id);
    return true;
  }
}
